ClickStack - Nuture flow - Otel collector (#6335)
ClickHouse/clickhouse-docs: docs/use-cases/observability/clickstack/managed-onboarding/setting-up-your-opentelemetry-collector.md

<VerticalStepper headerLevel="h2">

## Gather your credentials {#gather-credentials}

You'll need:

- The HTTPS endpoint of your ClickHouse Cloud service, including protocol and port, for example `https://abc123xyz.us-central1.gcp.clickhouse.cloud:8443`.
- A ClickHouse username and password for ingestion.

If you don't have these recorded, open your service in the [ClickHouse Cloud console](https://console.clickhouse.cloud) and select **Connect**. Record the url from the subsequent dialog. We will create a dedicated user for ingestion below.

<Image img={clickhouse_cloud_connection} size="lg" alt="Service connect panel showing HTTPS endpoint and password" border/>

## Create an ingestion user {#create-ingestion-user}

We recommend creating a dedicated user for the collector rather than reusing `default`. Connect to your service via the SQL console and run:

```sql
CREATE USER hyperdx_ingest IDENTIFIED WITH sha256_password BY 'ClickH0u3eRocks123!';
GRANT SELECT, INSERT, CREATE DATABASE, CREATE TABLE, CREATE VIEW ON otel.* TO hyperdx_ingest;
```

> **Replace the password with a strong value**. 

Reviewer: ```suggestion
:::tip
Replace the password in the snippet above with a strong value
:::
```

The collector creates the schema for logs, traces, and metrics inside the `otel` database on first use. For more guidance on production user setup, see [Going to production](/use-cases/observability/clickstack/production#create-a-database-ingestion-user-managed).

## Deploy the collector {#deploy-the-collector}

> Deploy the collector somewhere that is accessible to the applications and infrastructure sending OpenTelemetry data. In the example below, we run the collector locally and generate artificial telemetry from the same machine for simplicity.

Reviewer: ```suggestion
Deploy the collector somewhere that's accessible to the applications and infrastructure sending OpenTelemetry data. In the example below, we run the collector locally and generate artificial telemetry from the same machine for simplicity.
```

:::

Pick a shared secret to authenticate clients sending data to the collector, then export it alongside your connection details and chosen password for the `hyperdx_ingest` user:

```shell
export CLICKHOUSE_ENDPOINT=<HTTPS_ENDPOINT>
export CLICKHOUSE_USER=hyperdx_ingest
export CLICKHOUSE_PASSWORD=ClickH0u3eRocks123!
export OTLP_AUTH_TOKEN="a-strong-shared-secret"
```

Run the ClickStack OTel collector:

```shell
docker run -d \
  -e OTLP_AUTH_TOKEN=${OTLP_AUTH_TOKEN} \
  -e CLICKHOUSE_ENDPOINT=${CLICKHOUSE_ENDPOINT} \
  -e CLICKHOUSE_USER=${CLICKHOUSE_USER} \
  -e CLICKHOUSE_PASSWORD=${CLICKHOUSE_PASSWORD} \
  -e HYPERDX_OTEL_EXPORTER_CLICKHOUSE_DATABASE=otel \
  -p 4317:4317 \
  -p 4318:4318 \
  clickhouse/clickstack-otel-collector:latest
```

The collector now exposes OTLP gRPC on `4317` and OTLP HTTP on `4318`. Applications, SDKs, and agent collectors should send to these ports with `authorization: $OTLP_AUTH_TOKEN` in the request headers.

> :::note Production deployments

Reviewer: ```suggestion
:::note[Production deployments]
```

For production, we recommend enabling TLS on the OTLP endpoint. See [Securing the collector](/use-cases/observability/clickstack/ingesting-data/otel-collector#securing-the-collector).
:::

## Verify the endpoint {#verify-the-endpoint}

Generate some synthetic traffic against the collector to confirm the full pipeline works. We use [`otelgen`](https://github.com/krzko/otelgen), a small CLI that emits OTLP logs, traces, and metrics.

Install `otelgen` with Homebrew:

```shell
brew install krzko/tap/otelgen
```

Or with Go:

```shell
go install github.com/krzko/otelgen@latest
```

Send a short burst of logs to the collector:

```shell
 otelgen \
  --otel-exporter-otlp-endpoint localhost:4317 \
  --insecure \
  --protocol grpc \
  --header "authorization=${OTLP_AUTH_TOKEN}" \
  --rate 5 \
  --duration 60 \
  logs multi
```

For the equivalent trace and metrics commands, and a walkthrough of the other `otelgen` subcommands, see [Synthetic data with otelgen](/use-cases/observability/clickstack/getting-started/otelgen).

## Confirm in the ClickStack UI {#confirm-in-ui}

> Open your service in the [ClickHouse Cloud console](https://console.clickhouse.cloud) and select **ClickStack** from the left menu. 

Reviewer: ```suggestion
Open your service in the [ClickHouse Cloud console](https://console.clickhouse.cloud) and select **ClickStack** from the left menu.
```
